server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2687,4 +2665,24 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
             }
         };
     }
+
+    /**
+     * Rollback the current engine to the safe commit, then replay local translog up to the global checkpoint.
+     */
+    void resetEngineToGlobalCheckpoint() throws IOException {
+        assert getActiveOperationsCount() == 0 : "Ongoing writes [" + getActiveOperations() + "]";
+        sync(); // persist the global checkpoint to disk
+        final long globalCheckpoint = getGlobalCheckpoint();
+        final long maxSeqNo = seqNoStats().getMaxSeqNo();
+        logger.info("resetting replica engine from max_seq_no [{}] to global checkpoint [{}]", maxSeqNo, globalCheckpoint);
+        final Engine resettingEngine;
+        synchronized (mutex) {
+            IOUtils.close(currentEngineReference.getAndSet(null));
+            resettingEngine = createNewEngine(newEngineConfig());
+            currentEngineReference.set(resettingEngine);
+        }
+        final Engine.TranslogRecoveryRunner translogRunner = (engine, snapshot) ->
```

Comment:
it will be good to have some kind of progress logs here (like log ever 10k ops or something) under debug.

Reply:
We intend to have a dedicate recovery stats for the reset.